lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php

```diff
@@ -0,0 +1,311 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\KeyValueStore\Storage;
+
+use Aws\DynamoDb\DynamoDbClient;
+use Aws\DynamoDb\Marshaler;
+use Doctrine\KeyValueStore\NotFoundException;
+
+class AmazonDynamoDbStorage implements Storage
+{
+    /**
+     * @var DynamoDbClient
+     */
+    private $client;
+
+    /**
+     * @var Marshaler
+     */
+    private $marshaler;
+
+    /**
+     * @var string
+     */
+    private $defaultKeyName = 'Id';
+
+    /**
+     * A associative array where the key is the table name and the value is the name of the key.
+     *
+     * @var array
+     */
+    private $tableKeys = [];
+
+    /**
+     * @param DynamoDbClient $client    The client for connecting to AWS DynamoDB.
+     * @param Marshaler|null $marshaler (optional) Marshaller for converting data to/from DynamoDB format.
+     * @param string         $defaultKeyName   (optional) Default name to use for keys.
+     * @param array $tableKeys $tableKeys (optional) An associative array for keys representing table names and values
+     * representing key names for those tables.
+     */
+    public function __construct(
+        DynamoDbClient $client,
+        Marshaler $marshaler = null,
+        $defaultKeyName = null,
+        array $tableKeys = []
+    ) {
+        $this->client = $client;
+        $this->marshaler = $marshaler ?: new Marshaler();
+
+        if ($defaultKeyName !== null) {
+            $this->setDefaultKeyName($defaultKeyName);
+        }
+
+        if (!empty($tableKeys)) {
+            foreach ($tableKeys as $table => $keyName) {
+                $this->setKeyForTable($table, $keyName);
+            }
+        }
+    }
+
+    /**
+     * Validates a DynamoDB key name.
+     *
+     * @param $name mixed The name to validate.
+     *
+     * @throws \InvalidArgumentException When the key name is invalid.
+     */
+    protected function validateKeyName($name)
+    {
+        if (!is_string($name)) {
+            throw new \InvalidArgumentException(
```

Comment:
Please throw specialized exceptions, not SPL exceptions (write a new exception class - you may look at http://rosstuck.com/formatting-exception-messages/ )